src/onion.rs

```diff
@@ -267,40 +292,49 @@ impl From<ser::Error> for OnionError {
 #[cfg(test)]
 pub mod test_util {
 	use super::{Onion, OnionError, RawBytes};
-	use crate::secp::test_util::{rand_commit, rand_proof, rand_pubkey};
-	use crate::secp::{self, Commitment, PublicKey, Secp256k1, SecretKey, SharedSecret};
+	use crate::secp::test_util::{rand_commit, rand_proof};
+	use crate::secp::{random_secret, Commitment, SecretKey};
 	use crate::types::Payload;
 
 	use chacha20::cipher::StreamCipher;
 	use grin_core::core::FeeFields;
-	use rand::RngCore;
+	use rand::{thread_rng, RngCore};
+	use x25519_dalek::PublicKey as xPublicKey;
+	use x25519_dalek::{SharedSecret, StaticSecret};
 
 	#[derive(Clone)]
 	pub struct Hop {
-		pub pubkey: PublicKey,
+		pub pubkey: xPublicKey,
 		pub payload: Payload,
 	}
+	/*
+	Choose random xi for each node ni and create a Payload (Pi) for each containing xi
+	Build a rangeproof for Cn=Cin+(Σx1...n)*G and include it in payload Pn
+	Choose random initial ephemeral keypair (r1, R1)
+	Derive remaining ephemeral keypairs such that ri+1=ri*Sha256(Ri||si) where si=ECDH(Ri, Ki)

```

Comment:
Is this comment `Derive remaining ephemeral keypairs such that ri+1=ri*Sha256(Ri||si) where si=ECDH(Ri, Ki)` correct?

On line 333 we have
```
ephemeral_key = StaticSecret::from(
				*ephemeral_key
					.diffie_hellman(&xPublicKey::from(&blinding_factor))
					.as_bytes(),
			);
```

If I understand correctly, ephemeral_key is `ri` and this line defines `ri+1` as `StaticSecret::from((ri * xPublicKey::from(sha256(Ri || si))).as_bytes()` meaning we compute the shared secret, cast it to bytes and derive a secret from that random curve point? I don't see why `ri+1 = ri * sha256(Ri || si)` would hold here as we get random bytes, no?

Reply:
What is the reason for making each r_i dependent on the earlier one? Why not choose all at random?

Reply:
@phyro You're correct, that comment is no longer accurate. I'll get it fixed.

@tromp This reduces the size of the Onion packet. Only one pubkey needs to be sent, rather than 1 per payload. It's probably overkill though. I put a lot of time and effort into the scheme, and now I'm really regretting it lol. I'll switch to using random public keys in a separate PR.

Reply:
Issue created: https://github.com/mimblewimble/mwixnet/issues/19